src/operator/tensor/sparse_retain-inl.h

```diff
@@ -0,0 +1,315 @@
+ * in the in_idx, one should use SparseRetainRspRowBlockKernel instead,
+ * where each thread only perform binary search once.
+ */
+struct SparseRetainRspThreadKernel {
+  template<typename DType, typename RType, typename IType>
+  MSHADOW_XINLINE static void Map(int i, DType* out_data, RType* out_idx,
+                                  const DType* in_data, const RType* in_idx,
+                                  const IType* idx, const size_t nnr,
+                                  const size_t row_length) {
+    const RType irow = idx[i];
+    int j = -1, left = 0, right = nnr - 1;
+    while (left <= right) {
+      int m = left + (right - left) / 2;
+      const auto in_idx_m = in_idx[m];
+      if (in_idx_m == irow) {
+        j = m;
+        break;
+      } else if (in_idx_m < irow) {
+        left = m + 1;
+      } else {
+        right = m - 1;
+      }
+    }
+    out_idx[i] = idx[i];
+    if (j >= 0) {
+      const size_t in_offset = j * row_length;
+      const size_t out_offset = i * row_length;
+      for (size_t k = 0; k < row_length; ++k) {
+        out_data[out_offset+k] = in_data[in_offset+k];
+      }
+    }
+  }
+};
+
+/*!
+ * \brief This kernel is invoked when the input row-sparse
+ * is actually dense.
+ * Each thread searches for a subarray of indices of
+ * the user-input idx array for retain. The first index
+ * in the subarray will be searched for using binary search.
+ * The rest of the indices will be searched for starting from
+ * the lower bound of the binary search. This kernel assumes
+ * that idx has been sorted in ascending order.
+ */
+struct SparseRetainRspRowBlockKernel {
+  template<typename DType, typename RType, typename IType>
+  MSHADOW_XINLINE static void Map(int i, DType* out_data, RType* out_idx,
+                                  const DType* in_data, const RType* in_idx,
+                                  const IType* idx, const size_t num_indices,
+                                  const size_t nnr, const size_t row_length,
+                                  const size_t seg_len) {
+    const size_t seg_start = i * seg_len;
+    if (seg_start >= num_indices) return;
+    const size_t seg_end = (seg_start+seg_len < num_indices? seg_start+seg_len : num_indices);
+    for (size_t j = seg_start; j < seg_end; ++j) {
+      out_idx[j] = idx[j];
+    }
+    // use binary search to find the lower bound of idx[seg_start] in in_idx
+    const RType* first = in_idx;
+    const RType* last = in_idx + nnr;
+    const auto val = idx[seg_start];
+    const RType* it;
+    int count = last - first, step;
+    while (count > 0) {
+      it = first;
+      step = count / 2;
+      it += step;
+      if (*it < val) {
+        first = ++it;
+        count -= step + 1;
+      } else {
+        count = step;
+      }
+    }
+    size_t cur_row_idx = first - in_idx;
+    // end of binary search
+    if (cur_row_idx == nnr ||  in_idx[cur_row_idx] > idx[seg_end-1]) {
```

Comment:
In the future we probably want to check the value of nnr and see whether linear scan or binary search is faster.